src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,525 @@
+}
+
+int32_t FileMetaData::version() const {
+  return impl_->version();
+}
+
+const std::string& FileMetaData::created_by() const {
+  return impl_->created_by();
+}
+
+int FileMetaData::num_schema_elements() const {
+  return impl_->num_schema_elements();
+}
+
+const SchemaDescriptor* FileMetaData::schema() const {
+  return impl_->schema();
+}
+
+void FileMetaData::WriteTo(OutputStream* dst) {
+  return impl_->WriteTo(dst);
+}
+
+// MetaData Builders
+// row-group metadata
+class ColumnMetaDataBuilder::ColumnMetaDataBuilderImpl {
+ public:
+  explicit ColumnMetaDataBuilderImpl(std::shared_ptr<WriterProperties> props,
+      const ColumnDescriptor* column, uint8_t* contents)
+      : properties_(props), column_(column) {
+    column_chunk_ = reinterpret_cast<format::ColumnChunk*>(contents);
+    column_chunk_->meta_data.__set_type(ToThrift(column->physical_type()));
+    column_chunk_->meta_data.__set_path_in_schema(column->path()->ToDotVector());
+    column_chunk_->meta_data.__set_codec(
+        ToThrift(properties_->compression(column->path())));
+    std::vector<format::Encoding::type> thrift_encodings;
+    std::set<Encoding::type> encodings_set;
+    ColumnEncodingSet encodings = properties_->encodings(column->path());
+    encodings_set.insert(encodings.def_level_encoding);
+    encodings_set.insert(encodings.rep_level_encoding);
+    if (properties_->dictionary_enabled()) {
+      encodings_set.insert(encodings.dictionary_encoding);
+    }
+    encodings_set.insert(encodings.value_encoding);
+    for (auto encoding : encodings_set) {
+      thrift_encodings.push_back(ToThrift(encoding));
+    }
+    column_chunk_->meta_data.__set_encodings(thrift_encodings);
+  }
+  ~ColumnMetaDataBuilderImpl() {}
+
+  // column chunk
+  void set_file_path(const std::string& val) { column_chunk_->__set_file_path(val); }
+
+  // column metadata
+  // Needed to override the original values when fallback encoding is used
+  void SetEncodingsWithDictionaryFallback() {
+    std::vector<format::Encoding::type> thrift_encodings;
+    std::set<Encoding::type> encodings_set;
+    ColumnEncodingSet encodings = properties_->encodings(column_->path());
+    encodings_set.insert(encodings.def_level_encoding);
+    encodings_set.insert(encodings.rep_level_encoding);
+    if (properties_->dictionary_enabled()) {
+      encodings_set.insert(encodings.dictionary_encoding);
+    }
+    encodings_set.insert(encodings.value_encoding);
+    encodings_set.insert(encodings.dictionary_fallback_encoding);
+    for (auto encoding : encodings_set) {
+      thrift_encodings.push_back(ToThrift(encoding));
+    }
+    column_chunk_->meta_data.__set_encodings(thrift_encodings);
+  }
+
+  void SetStatistics(const ColumnStatistics& val) {
+    format::Statistics stats;
+    stats.null_count = val.null_count;
+    stats.distinct_count = val.distinct_count;
+    stats.max = *val.max;
+    stats.min = *val.min;
+
+    column_chunk_->meta_data.__set_statistics(stats);
```

Comment:
This results in the column stats strings being copied twice. If you set the `__isset.statistics` flag manually you can edit in place